// naive-ui/lib/tree/src/TreeNodeContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const _utils_1 = require("../../_utils");
const interface_1 = require("./interface");
exports.default = vue_1.defineComponent({
    name: 'TreeNodeContent',
    props: {
        clsPrefix: {
            type: String,
            required: true
        },
        disabled: {
            type: Boolean,
            default: false
        },
        checked: Boolean,
        selected: Boolean,
        onClick: Function,
        onDragstart: Function,
        tmNode: {
            type: Object,
            required: true
        }
    },
    setup(props) {
        const { renderLabelRef, renderPrefixRef, renderSuffixRef } = 
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        vue_1.inject(interface_1.treeInjectionKey);
        const selfRef = vue_1.ref(null);
        function doClick(e) {
            const { onClick } = props;
            if (onClick)
                onClick(e);
        }
        function handleClick(e) {
            doClick(e);
        }
        return {
            selfRef,
            renderLabel: renderLabelRef,
            renderPrefix: renderPrefixRef,
            renderSuffix: renderSuffixRef,
            handleClick
        };
    },
    render() {
        const { clsPrefix, checked = false, selected = false, renderLabel, renderPrefix, renderSuffix, handleClick, onDragstart, tmNode: { rawNode, rawNode: { prefix, label, suffix } } } = this;
        return (vue_1.h("span", { ref: "selfRef", class: [`${clsPrefix}-tree-node-content`], onClick: handleClick, draggable: onDragstart === undefined ? undefined : true, onDragstart: onDragstart },
            renderPrefix || prefix ? (vue_1.h("div", { class: `${clsPrefix}-tree-node-content__prefix` }, renderPrefix
                ? renderPrefix({
                    option: rawNode,
                    selected,
                    checked
                })
                : _utils_1.render(prefix))) : null,
            vue_1.h("div", { class: `${clsPrefix}-tree-node-content__text` }, renderLabel
                ? renderLabel({
                    option: rawNode,
                    selected,
                    checked
                })
                : _utils_1.render(label)),
            renderSuffix || suffix ? (vue_1.h("div", { class: `${clsPrefix}-tree-node-content__suffix` }, renderSuffix
                ? renderSuffix({
                    option: rawNode,
                    selected,
                    checked
                })
                : _utils_1.render(suffix))) : null));
    }
});
