// naive-ui/lib/rate/src/styles/index.cssr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr"); // vars:
// --bezier
// --item-color
// --item-color-active


exports.default = cssr_1.cB('rate', {
  display: 'inline-flex',
  flexWrap: 'nowrap'
}, [cssr_1.c('&:hover', [cssr_1.cE('item', `
 transition:
 transform .1s var(--bezier),
 color .3s var(--bezier);
 `)]), cssr_1.cE('item', `
 position: relative;
 display: flex;
 transition:
 transform .1s var(--bezier),
 color .3s var(--bezier);
 transform: scale(1);
 font-size: var(--item-size);
 color: var(--item-color);
 `, [cssr_1.c('&:not(:first-child)', {
  marginLeft: '6px'
}), cssr_1.cM('active', {
  color: 'var(--item-color-active)'
})]), cssr_1.cNotM('readonly', `
 cursor: pointer;
 `, [cssr_1.cE('item', [cssr_1.c('&:hover', {
  transform: 'scale(1.05)'
}), cssr_1.c('&:active', {
  transform: 'scale(0.96)'
})])]), cssr_1.cE('half', `
 display: flex;
 transition: inherit;
 position: absolute;
 top: 0;
 left: 0;
 bottom: 0;
 width: 50%;
 overflow: hidden;
 color: var(--item-color);
 `, [cssr_1.cM('active', {
  color: 'var(--item-color-active)'
})])]);
